keps/sig-scheduling/5501-reflect-preenqueue-rejections-in-pod-status/README.md

```diff
@@ -0,0 +1,705 @@
+- Provide a mechanism for existing `PreEnqueue` use cases (like `DynamicResources` plugin)
+  to adopt this new status reporting.
+- Investigate migrating the responsibility for setting the `SchedulingGated` condition
+  from the kube-apiserver to the kube-scheduler for consistency.
+
+### Non-Goals
+
+- This KEP will not change the core rejection logic of any existing `PreEnqueue` plugin,
+  only how the outcome of that logic is communicated.
+- This KEP will not alter the scheduling cycle or the behavior of other scheduler extension points.
+
+## Proposal
+
+When a `PreEnqueue` plugin rejects a Pod, the scheduler will asynchronously patch the Pod's status
+to reflect this outcome. The patch will add or update a `PodCondition` to indicate that the Pod is not schedulable
+due to a `PreEnqueue` check. To prevent redundant API calls, the scheduler will internally cache the last-reported status
+and skip the patch if the rejection reason has not changed.
+
+**Note:** This entire feature will be controlled by a new feature gate, `SchedulerPreEnqueuePodStatus`.
+The enablement of this feature will be dependent on the `SchedulerAsyncAPICalls` feature gate being enabled,
+as the asynchronous patching mechanism is critical to protect scheduler throughput.
+
+### 1. How should plugins provide the status message?
+
+**Proposed:** Adjust the `PreEnqueue` plugin interface to return an additional,
+optional, user-facing message alongside the `Status` object.
+This will introduce a **breaking change** to all plugins that implement the `PreEnqueue` extension point.
+
+This flexibility is important for plugins that wish to avoid reporting transient rejections.
+For example, the `DynamicResources` plugin might observe a rejection because the scheduler processes a Pod faster
+than a `ResourceClaim` becomes visible through the watcher (see a [comment](https://issues.k8s.io/129698#issuecomment-2614991955)).
+In such cases where the condition is expected to resolve in seconds, populating the status would be inappropriate noise.
+
+- **Pros:**
+  - Gives plugin authors explicit control over what users see.
+  - Decouples internal rejection reasons from user-facing messages.
+  - Allows plugins to opt out of reporting by returning an empty message.
+
+- **Cons:**
+  - Requires a breaking change to the `PreEnqueue` plugin interface.
+
+- **Alternatives Considered:**
+  - Use the raw status message: This is simpler as it requires no interface changes.
+    However, these messages are often not written well for end-users.
+    It also makes it difficult for a plugin to conditionally opt out of reporting a status.
+
+### 2. Should all PreEnqueue plugins report the status?
+
+**Proposed:** Make status reporting an opt-in or easily controllable behavior at the plugin level
+(e.g., by returning an empty message described above).
+
+- **Pros:**
+  - Reduces API server load and potential "noise" for plugins that reject pods for very short,
```

Comment:
To mitigate it, can we just skip an error reporting if that's identical with the existing one?
e.g., if the pod keeps being rejected by the preemption PreEnqueue plugin, it will get a message like `This pod is not schedulable because waiting for the preemption to be completed` for the first time, but the preemption PreEnqueue plugin will (i.e. should) return the same common message and then the scheduling framework doesn't update the pod with the message every time PreEnqueue rejects the same pod with the same reason.